inference-engine/src/transformations/src/transformations/common_optimizations/algebraic_simplification.cpp

```diff
@@ -234,31 +235,32 @@ static bool replace_transpose_with_reshape(shared_ptr<Node> transpose) {
     return true;
 }
 
-bool pass::AlgebraicSimplification::run_on_function(shared_ptr<Function> f) {
-    RUN_ON_FUNCTION_SCOPE(AlgebraicSimplification);
-    static const unordered_map<NodeTypeInfo, function<bool(shared_ptr<Node>)>> ops_to_simplifiers =
-        {{opset3::Gather::type_info, simplify_gather},
-         {opset2::ShapeOf::type_info, simplify_gather_shapeof},
-         {opset3::ShapeOf::type_info, simplify_gather_shapeof},
-         {opset3::Transpose::type_info, replace_transpose_with_reshape}};
-
-    bool replaced = false;
-    for (auto n : f->get_ordered_ops()) {
-        if (op::is_output(n) || op::is_parameter(n)) {
-            continue;
-        }
-
-        // Recursively apply transformation for sub-graph based operations
-        if (auto sub_graph_node = std::dynamic_pointer_cast<op::util::SubGraphOp>(n)) {
-            if (auto sub_graph = sub_graph_node->get_function()) {
-                replaced |= run_on_function(sub_graph);
-            }
-        }
-
-        auto eh = ops_to_simplifiers.find(n->get_type_info());
-        if (eh != ops_to_simplifiers.end()) {
-            replaced |= eh->second(n);
-        }
-    }
-    return replaced;
+#define ECHO(NAME) #NAME
+#define STR(NAME) ECHO(NAME)
+#define SIMPLE_MATCHER_PASS_DEFINITION(NAME, OP, FUNC) \
+class NAME : public ngraph::pass::MatcherPass { \
+public: \
+NGRAPH_RTTI_DECLARATION; \
+NAME() { \
+    MATCHER_SCOPE(NAME);    \
+    auto match_node = ngraph::pattern::wrap_type<OP>(); \
+    ngraph::matcher_pass_callback callback = [=](ngraph::pattern::Matcher &m) { \
+        return FUNC(m.get_match_root()); \
+    }; \
+    auto m = std::make_shared<ngraph::pattern::Matcher>(match_node, matcher_name); \
+    register_matcher(m, callback); \
+}  \
+}; \
```

Comment:
maybe we can use templates? E.g. create template class and instantiate with different ops and functions:
```c++
template <typename Op, typename Func>
class SimpleMatcher { .. };

ngraph::pass::AlgebraicSimplification::AlgebraicSimplification() {
    add_matcher<SimpleMatcher<opset3::Gather, simplify_gather>>();
    add_matcher<SimpleMatcher<opset2::ShapeOf, simplify_gather_shapeof>>();
    add_matcher<SimpleMatcher<opset3::ShapeOf, simplify_gather_shapeof>>();
    add_matcher<SimpleMatcher<opset3::Transpose, replace_transpose_with_reshape>>();
}
```

BTW, `Func` can be passed as a constructor argument.

Reply:
Anyway we will have to define NGRAPH_RTTI_DEFINITION for each specialization and also this passes (AlgebraicSimplification and NopElimination) will be refactored (Ticket: 37380) bun in a separate PR.

 The changes I've made are temporary but they are necessary to have this passes be united into single graph traversal.